tests/deploy/test_scheduler_agent.py

```diff
@@ -3137,12 +3138,199 @@ async def restore_baseline_state() -> None:
+    await scheduler._new_version(
+        [
+            model(
+                {
+                    rid1: make_resource_minimal(rid1, values={"changed": "unknown"}, requires=[]),
+                    rid2: make_resource_minimal(rid2, values={"changed": "unknown"}, requires=[]),
+                    rid3: make_resource_minimal(rid3, values={"changed": "unknown"}, requires=[]),
+                },
+                undefined={rid1, rid2, rid3},
+            ),
+            model(
+                {
+                    rid1: all_models[-1].resources[rid1],
+                    rid2: make_resource_minimal(rid2, values={"changed": "unknown"}, requires=[]),
+                    rid3: make_resource_minimal(rid3, values={"changed": "unknown"}, requires=[]),
+                },
+                undefined={rid2, rid3},
+            ),
+            model(
+                {
+                    rid1: all_models[-1].resources[rid1],
+                    rid2: all_models[-1].resources[rid2],
+                    rid3: make_resource_minimal(rid3, values={"changed": "unknown"}, requires=[]),
+                },
+                undefined={rid3},
+            ),
+            model(
+                {
+                    rid1: all_models[-1].resources[rid1],
+                    rid2: all_models[-1].resources[rid2],
+                    rid3: all_models[-1].resources[rid3],
+                },
+            ),
+        ]
+    )
+    assert scheduler._state.resources == all_models[-1].resources
+    for rid in (rid1, rid2, rid3):
+        assert scheduler._state.resource_state[rid].status is ComplianceStatus.HAS_UPDATE
+        assert scheduler._state.resource_state[rid].deployment_result is DeploymentResult.DEPLOYED
+        assert scheduler._state.resource_state[rid].blocked is BlockedStatus.NO
+
+    await restore_baseline_state()
+
+    # Resource becomes undefined, defined and then undefined again
+    await scheduler._new_version(
+        [
+            model(
+                {
+                    rid1: make_resource_minimal(rid1, values={"changed": "unknown"}, requires=[]),
+                },
+                undefined={rid1},
+            ),
+            model(
+                {
+                    rid1: all_models[-1].resources[rid1],
+                },
+            ),
+            model(
+                {
+                    rid1: make_resource_minimal(rid1, values={"changed": "unknown"}, requires=[]),
+                },
+                undefined={rid1},
+            ),
+        ]
+    )
+    assert scheduler._state.resources == all_models[-1].resources
+    assert scheduler._state.resource_state[rid1].status is ComplianceStatus.UNDEFINED
+    assert scheduler._state.resource_state[rid1].deployment_result is DeploymentResult.DEPLOYED
+    assert scheduler._state.resource_state[rid1].blocked is BlockedStatus.YES
+
+    await restore_baseline_state()
+
+    # Resource becomes defined, undefined and then defined again
```

Comment:
I like it, but I think this one requires a setup stage as well, because the resource doesn't "become defined" in the first version, it "remains defined" because `restore_baseline_state` makes it defined.